### ISSUE TITLE
Warning 32 ought to be respected

### ISSUE THREAD

**Issue by jberdine** [cristianoc/reanalyze#8]
Attributes that disable warning 32 in sections of source code are used to silence the compiler's dead code warnings. It would be best if reanalyze were to also respect these warnings. For example, `deriving compare` in signatures is careful to disable warning 32 to prevent reports on generated code:
```ocaml
let compare_int = compare

module M : sig
  type t = int [@@deriving_inline compare]

  include sig
    [@@@ocaml.warning "-32"]

    val compare : t -> t -> int
  end
  [@@ocaml.doc "@inline"]

  [@@@end]
end = struct
  type t = int [@@deriving_inline compare]

  let _ = fun (_ : t) -> ()
  let compare = (compare_int : t -> t -> int)
  let _ = compare

  [@@@end]
end
```
This is used in generated code, but also in non-generated code, such as [here](https://github.com/facebook/infer/blob/8d20e4d64db0118da5fe9a3fa1130c5b5bef18f7/infer/src/istd/ARList.mli#L66).

**Comment by chenglou** [comment]
Hey Josh. Unrelated but glad to see you again =)

**Comment by cristianoc** [comment]
I guess this falls in the category: it is dead, it should be considered dead, but should not be reported as dead.
Not sure I'd want to consider global options, only these local `@@@ocaml.warning "-32"`. Which makes it an ad-hoc treatment. Also, not clear how to get hold of global options anyway. So one would observe a difference when disabling `32` in the command line, and in the code.

**Comment by jberdine** [comment]
Yes: dead but shouldn't be reported. But it is not dead in the `[@live]` sense rather than the `[@dead]` one, since such code might be the only use of other code, and that other code should not be reported dead.

Considering only the attributes seems pragmatically sensible, since it isn't clear what the use case might be for compiling with `-w -32` and expecting dead code analysis to not be noisy.

FWIW, I don't think that this really has anything crucially to do with ppx. Warning 32 predates ppx, and there are reasonable cases where e.g. a codebase's style mandates that each type comes with a common set of operations like `compare` and `hash`.

Also, it is probably a good idea to handle the other "unused" warnings similarly. Maybe 32-39, 60, 66.

**Comment by jberdine** [comment]
edit: s/dead/not dead/

**Comment by cristianoc** [comment]
There’s also the automatic deletion of dead code. And for that one would like to delete even things kept alive only by this.
But it’s not a big issue. I’ll keep it simple.

**Comment by jberdine** [comment]
Ok, but for automatic deletion of dead code it's a ppx that outputs a new parsetree right, so no need to change the source, just give it to the rest of the compiler.

**Comment by cristianoc** [comment]
Currently the analysis adds annotations, and the deletion  ppx runs afterwards. So there’s a limit anyway to what can be done with that setup, as you can’t annotate code generated by a ppx.
A fuller solution would require deep integration in the compiler, as typed ast to typed ast xform.

**Comment by cristianoc** [comment]
So for now it seems that indeed treating 42 as @live is the way forward.

**Comment by jberdine** [comment]
> Currently the analysis adds annotations

Does this work by adding attributes to the parsetree? If so, it would be rolled into one ppx pipeline and not have to worry about whether code was generated or not.

**Comment by cristianoc** [comment]
No it’s literally printing into the source file.
So it’s very easy to take a quick look at a project without going through the list of reported issues.

**Comment by cristianoc** [comment]
Btw the analysis operates on the typed tree. Does not fit a ppx pipeline.

**Comment by jberdine** [comment]
Ah, yes, I knew but overlooked that. It would be a bit of work then to get back to the untyped tree, or construct a new one, or... Hmm. I suspect that you could reuse the implementation technique of ppx_import though, as it reads cmt files and generates code. That needs to use the staged_pps instead of pps flow in dune, but that should also be doable for reanalyze I think.

**Comment by cristianoc** [comment]
The analysis is global so strictly speaking does not fit in the ppx flow by that reason alone. However of course it could store the results somewhere then run a ppx that reads them.

**Comment by jberdine** [comment]
Oh yeah, very true. I imagine that you can figure out something compositional to do... ;-)

**Comment by cristianoc** [comment]
I’d have first to make it slower, to give an excuse.

**Comment by cristianoc** [comment]
@jberdine this is done in https://github.com/cristianoc/reanalyze/pull/13.
Unsure about whether to support other suppressions beyond `32`. What did you have in mind?

**Comment by cristianoc** [comment]
A relevant one is `34 Unused type declaration`. Though that's a bit different from the sense in which variant/records parts are unused.
Merging this for now.

**Comment by jberdine** [comment]
The following might also make sense:
 37 Unused constructor.
 38 Unused extension constructor.
 60 Unused module declaration.

**Comment by jberdine** [comment]
It seems that code like
```
let[@warning "-32"] foo x = print_int x
```
still gets flagged if `foo` is otherwise dead.

**Comment by cristianoc** [comment]
Let me check.

**Comment by cristianoc** [comment]
Tried to repro the example here:
https://github.com/cristianoc/reanalyze/pull/22

No cigar.

What was the example?

**Comment by cristianoc** [comment]
Oh wait `warning` not `ocaml.warning`?

**Comment by jberdine** [comment]
Yes, they are synonyms.
